### PR TITLE
Features/viewpoints

### PR DESCRIPTION
**Pull request by guylerme**
This is the first version of the implementation of Viewpoint Extraction

1: Implementtion of Kind Viewpoint. This viewpoint group all kinds and relations between then

2: Implementation of Subkind Viewpoint. This viewpoint group all subkinds, sortal lines and relations between then

3: Implementation of Role Viewpoint. This viewpoint group all role, sortal lines and relations between then

4: Implementation of Phase Viewpoint. This viewpoint group all phase, sortal lines and relations between then

5: Implementation of Relator Viewpoint. This viewpoint group all relators, classes where it mediates, sortal lines and relations between then

6: Implementation of Non-Sortal Viewpoint. This viewpoint group all non-sortals, non-sortals decendants, decendants ancestors sortal lines and relations between then

7: Implementation of Mode Viewpoint. This viewpoint group all modes, classes characterized by it, sortal lines and relations between then

All test cases were run in "__tests __ / lib / viewpoints ". The tests verify the elements in model after execute the rule.

### CODE REVIEW

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+     * 
+     * Step 3 Get all relations between all classes (subkinds and sortal ancestors)
+     * 
+     * Step 4 Get all generalizations between all classes (subkinds and sortal ancestors)
+     */
+
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.SUBKIND);
+
+    let generals = [];
+    for (let i = 0; i < classes.length; i++) {
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:122]: `Class` has a method called `getSortalAncestors()` which seem to do exactly what you did here. Please check.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+    /**Step 1 Get all Subkinds in model
+     * 
+     * Step 2 Get all sortal ancestors 
+     * 
+     * Step 3 Get all relations between all classes (subkinds and sortal ancestors)
+     * 
+     * Step 4 Get all generalizations between all classes (subkinds and sortal ancestors)
+     */
+
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.SUBKIND);
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:119]: `Project` has a method called `getClassesWithSubkindStereotype()`, please check.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+  kindView(): Viewpoint {
+    const viewpoint = new Viewpoint('Kind View');
+
+    /**Step 1 Get all Kinds in model
+     * 
+     * Step 2 Get all relations between Kinds
+     * 
+     * Step 3 Get all generalizations between Kinds
+     */
+
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:90]: `Project` has a method `getClassesWithKindStereotype()`. Please check and reuse it.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+    viewpoint.addClasses(classes);
+    viewpoint.addClasses(generals);
+
+    viewpoint.removeDuplicates();
+
+    viewpoint.addRelations(this.getRelationsBetweenClasses(viewpoint.classes));
+    viewpoint.addGeneralizations(this.getGeneralizationsBetweenClasses(viewpoint.classes));
+
+    viewpoint.removeDuplicates();
+
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:135]: Aren't generalization sets included in the subkind view?

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+    for (let i = 0; i < classes.length; i++) {
+      generals = generals.concat(classes[i].getAncestors().filter((cl) => cl.hasSortalStereotype()));
+    }
+
+    viewpoint.addClasses(classes);
+    viewpoint.addClasses(generals);
+
+    viewpoint.removeDuplicates();
+
+    viewpoint.addRelations(this.getRelationsBetweenClasses(viewpoint.classes));
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:131]: This does not seem to be the view proposed in the paper.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+    /**Step 1 Get all Roles in model
+     * 
+     * Step 2 Get all sortal ancestors 
+     * 
+     * Step 3 Get all relations between all classes (roles and sortal ancestors)
+     * 
+     * Step 4 Get all generalizations between all classes (roles and sortal ancestors)
+     */
+
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.ROLE);
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:157]: Replace with appropriate method.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+
+    let generals = [];
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:159]: `Class` has a method called `getSortalAncestors()`

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+     * 
+     * Step 2 Get all sortal ancestors 
+     * 
+     * Step 3 Get all relations between all classes (phases and sortal ancestors)
+     * 
+     * Step 4 Get all generalizations between all classes (phases and sortal ancestors)
+     */
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.PHASE);
+
+    let generals = [];
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:196]: `Class` has a method called `getSortalAncestors()`

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+     * 
+     * Step 2 Get all classes conected by mediation relation 
+     * 
+     * Step 3 Get all sortal ancestors 
+     * 
+     * Step 4 Get all relations between all classes 
+     * 
+     * Step 5 Get all generalizations between all classes 
+     */
+    let classes = this.project.getAllClassesByStereotype(ClassStereotype.RELATOR);
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:234]: Please use english names for variables and methods.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+     * Step 3 Get all decendants sortal ancestors 
+     * 
+     * Step 4 Get all relations between all classes 
+     * 
+     */
+    let classes = this.project.getAllClassesByStereotype(ClassStereotype.ROLE_MIXIN);
+
+    classes = classes.concat(this.project.getAllClassesByStereotype(ClassStereotype.MIXIN));
+    classes = classes.concat(this.project.getAllClassesByStereotype(ClassStereotype.CATEGORY));
+    classes = classes.concat(this.project.getAllClassesByStereotype(ClassStereotype.PHASE_MIXIN));
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:292]: Please add a method to the `Project` class called `getNonSortals()` that returns classes with these 4 stereotypes.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+     * Step 4 Get all relations between all classes 
+     * 
+     * Step 5 Get all generalizations between all classes 
+     */
+    let classes = this.project.getAllClassesByStereotype(ClassStereotype.RELATOR);
+
+    const relacoes = this.project.getAllRelations();
+    let mediacoes = [];
+
+    for (let j = 0; j < classes.length; j++) {
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:239]: Prefer using a for-in loop (https://www.w3schools.com/js/js_loop_forin.asp). Please check other usages.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+    }
+
+    viewpoint.addClasses(classes);
+    viewpoint.addClasses(generals);
+
+    viewpoint.removeDuplicates();
+
+    viewpoint.addRelations(this.getRelationsBetweenClasses(viewpoint.classes));
+    viewpoint.addGeneralizations(this.getGeneralizationsBetweenClasses(viewpoint.classes));
+
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:208]: Again, no generalization sets?

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 13]
```diff
@@ -0,0 +1,536 @@
...
+     * 
+     * Step 3 Get all sortal ancestors 
+     * 
+     * Step 4 Get all relations between all classes 
+     * 
+     * Step 5 Get all generalizations between all classes 
+     */
+    let classes = this.project.getAllClassesByStereotype(ClassStereotype.RELATOR);
+
+    const relacoes = this.project.getAllRelations();
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:236]: Please add a `getMediations()` method to the `Project` class.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 14]
```diff
@@ -0,0 +1,536 @@
...
+
+    return viewpoint;
+  }
+
+  /**
+   * Verify is a class participates in relation
+   * @param _class The classe to be verified
+   * @param relation The relation where the class may be participating
+   */
+  isClassInRelation(_class: Class, relation: Relation): boolean {
```
Comment (tgoprince) [inline src/libs/viewpoints/viewpoint_extractor.ts:371]: This should be a method of the `Relation` class named `involves()`

**Review by tgoprince — CHANGES_REQUESTED** [review]

File: __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+    });
+
+    it('Should contain ancestors of phase class: «phase» Temporarily Closed Harbor', () => {
+      expect(diagram.findView(k_harbor)).toBeTruthy();
+    });
+  });
+
+
+  describe('Role View', () => {
+    let diagram: Diagram = diagrams.find(d => d.getName() === 'Role View');
```
[the remark this answers is not in the document]
Comment (tgoprince) [reply __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts:179]: Including boundary cases and unhappy paths.

File: src/libs/viewpoints/viewpoint.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  GeneralizationSet,
+  Generalization,
+  Relation,
+  ModelElement,
+  Diagram,
+} from '@libs/ontouml';
+import {uniqBy} from 'lodash';
+
+
+export class Viewpoint {
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint.ts:13]: I agree. I will add viewpoint_extractor into complexity package.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 17]
```diff
@@ -0,0 +1,536 @@
...
+  }
+
+  buildAll(): Diagram[] {
+    const allPatterns = [ClassStereotype.KIND, ClassStereotype.SUBKIND, ClassStereotype.ROLE, ClassStereotype.PHASE, ClassStereotype.RELATOR, ClassStereotype.MODE, 'NON_SORTAL'];
+    return allPatterns.map((pattern, index) => this.extractViewpoint(String(index), pattern))
+      .map((viewpoint) => viewpoint.createDiagram(this.project.model));
+  }
+
+  /**
+   * This methor will run the patter viewpoint extrator.
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:46]: ok!

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 18]
```diff
@@ -0,0 +1,536 @@
...
+    this.project.addDiagrams(generatedDiagrams);
+
+    return {
+      result: this.project,
+      issues: null,
+    };
+  }
+
+  buildAll(): Diagram[] {
+    const allPatterns = [ClassStereotype.KIND, ClassStereotype.SUBKIND, ClassStereotype.ROLE, ClassStereotype.PHASE, ClassStereotype.RELATOR, ClassStereotype.MODE, 'NON_SORTAL'];
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:40]: Yes, but the method returns module, not diagram. I can make one per view to return diagram.

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 19]
```diff
@@ -0,0 +1,536 @@
...
+   */
+  kindView(): Viewpoint {
+    const viewpoint = new Viewpoint('Kind View');
+
+    /**Step 1 Get all Kinds in model
+     * 
+     * Step 2 Get all relations between Kinds
+     * 
+     * Step 3 Get all generalizations between Kinds
+     */
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:89]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 3 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:90]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 20]
```diff
@@ -0,0 +1,536 @@
...
+
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.KIND);
+
+    viewpoint.addClasses(classes);
+    viewpoint.addRelations(this.getRelationsBetweenClasses(classes));
+    viewpoint.addGeneralizations(this.getGeneralizationsBetweenClasses(classes));
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:95]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 21]
```diff
@@ -0,0 +1,536 @@
...
+    const viewpoint = new Viewpoint('Kind View');
+
+    /**Step 1 Get all Kinds in model
+     * 
+     * Step 2 Get all relations between Kinds
+     * 
+     * Step 3 Get all generalizations between Kinds
+     */
+
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.KIND);
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:91]: Kind view we use only Kinds

File: src/libs/viewpoints/viewpoint_extractor.ts [hunk 22]
```diff
@@ -0,0 +1,536 @@
...
+
+    viewpoint.addClasses(classes);
+    viewpoint.addRelations(this.getRelationsBetweenClasses(classes));
```
[the remark this answers is not in the document]
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:94]: Yes. Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 2 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:119]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 1 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:122]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 5 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:131]: Only Kind view doesn't have relations.

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 4 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:135]: We didn't specify this. At least the generalization must be. I dont't see problem considering generalizationset

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 6 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:157]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 12 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:208]: Can I put it?

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 9 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:234]: change classes?

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 10 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:292]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 14 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:371]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 13 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:236]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 11 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:239]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 8 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:196]: Done

File: src/libs/viewpoints/viewpoint_extractor.ts (hunk 7 above)
Comment (guylerme) [reply src/libs/viewpoints/viewpoint_extractor.ts:159]: Done

File: __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+  const rl_transportationContract = model.createRelator('TransportationContract')
+
+  const md_contract_administration = model.createMediationRelation(rl_transportationContract, rl_shipAdministration, 'md_contract_administration')
+  const md_contract_client = model.createMediationRelation(rl_transportationContract, rm_transportationContractClient, 'md_contract_client')
+
+  let diagrams: Diagram[] = new ViewpointExtractor(project).buildAll();
+
+  
+
+  describe('Kind View', () => {
```
[the remark this answers is not in the document]
Comment (guylerme) [reply __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts:108]: Done

File: __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts [hunk 24]
```diff
@@ -0,0 +1,366 @@
...
+    let diagram: Diagram = diagrams.find(d => d.getName() === 'Kind View');
+
+    it('Should contain the expected classes (4)', () => {
```
[the remark this answers is not in the document]
Comment (guylerme) [reply __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts:111]: Done, but need more

File: __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (guylerme) [reply __tests__/libs/viewpoints/viewpoint_extractor.basic.test.ts:179]: for additional tests I`m doing smaller models

**Review by tgoprince — APPROVED** [review]
